### DISCUSSION TITLE
Native: Static Map Rendering Optimization

### DISCUSSION THREAD

**Discussion by sjg-wdw** [maplibre/maplibre#334] [Native feature proposals]
## Static Map Renderering

Here's some functionality we'd like to add to MapLibre. This is one proposal of several for Native and if you'd like a little context and our strategy [there's an overview](https://www.wetdogweather.com/blog/maplibre-proposals). This one is purely Native.

Can you render static maps with MapLibre?  What even is a static map?  Is this an important use case?
- Yes.
- It's an old school image with all the map bits in it.
- Maybe.

Let's see what's wrong with the static map use case.

### Static Map Rendering

Put simply Static Map Rendering is the act of rendering a dynamic map with all its vectors and icons and images and text and whatever into an image and shipping that image off to be used.  These are often used in simpler web displays or on watches or places where you just what an image, not a ton of software.

This is how the [WMS GetMap](http://opengeospatial.github.io/e-learning/wms/text/operations.html#getmap) call works.  It's incredibly old school, but it's also still used in far more cases than you might think.  Optimizing for it is an afterthought.

### Rendering a Thing Once

MapLibre Native and JS are optimized to render map tiles a lot.  Ideally 60 times a second.  There is a lot of logic in the toolkit to set things up, consolidate things, and generally make things render fast.  It's all a waste for this use case.

Here the developer renders a view once.  Just once.  And then throws it all away and moves to an unrelated view.

Ideally, they can cache the style, so let's ignore that.  But for a given area the task looks like this:
- Figure out the tiles needed
- Load the tiles, probably asynchronously
- Turn each tile into renderable geometry
- Set up the shaders and uniforms and such
- Render the area in question
- Copy out and return the image
- Tear down the uniforms and state
- Tear down the tile geometry

That is a prodigious amount of setup and teardown work to do just one render.  Nothing in the toolkit is optimized for this use case.  It must be incredibly wasteful.

### Software vs. GPU Rendering

Could you improve the situation with GPU rendering?  Eh, probably not.

If you were making a video, maybe a flythrough or something, yes.  You'd be able to reuse all of the tiles and geometry and rendering infrastructure. But for a single frame, all that is a waste.

For modern GPU based rendering you do a lot of setup to efficiently draw a lot of frames.  The cost of that setup, copying data down, consolidating into optimized forms, setting yourself up to communicate efficiently with the shaders, it all pays off when you reuse most of it for the next 200 frames.  The cost of that communication is likely far larger than just doing something simple in software.

It's quite plausible that you're better off using a software renderer than a GPU based one for this case.  Even more likely when you factor in the cost of cloud-friendly GPUs.

### Optimizing for Software Rendering

What would it look like to optimize for software rendering?  For someone in real-time geospatial rendering, it's a fascinating question.

You might start by turning off all the geometry consolidation and optimization.  Putting several lines together to draw them once makes very little difference to the software renderer.

Then look at changeable values.  We know what the zoom level is and shall always be for the one frame we're going to render.  So just calculate that one value and stuff it in the proper uniform.

Is it practical to render as we parse?  That might be very difficult to keep compatible with the real-time renderer, but it's worth thinking about.

Tesselation is time consuming, as is parsing the vector tile and linking it up to the appropriate style.  Anything we can do less of there is good and our Optimized Map Pipeline proposal might fit in well here.

### Discussion

This is a weird use case specific to a few big users and a lot of smaller ones.  The costs can be enormous to the right user, but pretty mild to everyone else.

It's worth optimizing with the following criteria:
- Someone pays for it.  It needs to matter to them
- It doesn't interfere with the interactive use case in any significant way

What that looks like is up for discussion.  Well, not the money.  That's clear enough.  The bit about interfering with the interactive use case.  A few flags here and there are probably fine.  A complete restructuring to support software rendering, less so.

**Comment by wipfli** [comment]
Is this related to generating raster tiles with MapaLibre Native?

**Reply by sjg-wdw** [reply]
Not tiles, whole map areas.

On Apple's PhoneOS, for instance, it's pretty common to render a map for just the area the needs to see as a static image and use that.

I know Meta has a lot of use cases for this one.  Maybe Marc might want to weigh in.

**Reply by mwilsnd** [reply]
Meta's service renders both single tiles and static regions, though mostly tiles. Given the scale of the service, small improvements can really add up. The single biggest metric is end-to-end latency, starting from the time a fresh style is applied to when an image is returned from the frame buffer and encoded to a more usable image format.
